scipy/linalg/decomp.py

```diff
@@ -1267,3 +1267,138 @@ def hessenberg(a, calc_q=False, overwrite_a=False, check_finite=True):
+    stacked arrays.
+
+    Examples
+    --------
+    >>> import numpy
+    >>> X = numpy.array([[1, 2, 3], [0, 4, 5], [0, -5, 4]])
+    >>> X
+    array([[ 1,  2,  3],
+           [ 0,  4,  5],
+           [ 0, -5,  4]])
+
+    >>> from scipy import linalg
+    >>> w, v = linalg.eig(X)
+    >>> w
+    array([ 1.+0.j,  4.+5.j,  4.-5.j])
+    >>> v
+    array([[ 1.00000+0.j     , -0.01906-0.40016j, -0.01906+0.40016j],
+           [ 0.00000+0.j     ,  0.00000-0.64788j,  0.00000+0.64788j],
+           [ 0.00000+0.j     ,  0.64788+0.j     ,  0.64788-0.j     ]])
+
+    >>> wr, vr = linalg.cdf2rdf(w, v)
+    >>> wr
+    array([[ 1.,  0.,  0.],
+           [ 0.,  4.,  5.],
+           [ 0., -5.,  4.]])
+    >>> vr
+    array([[ 1.     ,  0.40016, -0.01906],
+           [ 0.     ,  0.64788,  0.     ],
+           [ 0.     ,  0.     ,  0.64788]])
+
+    >>> vr @ wr
+    array([[ 1.     ,  1.69593,  1.9246 ],
+           [ 0.     ,  2.59153,  3.23942],
+           [ 0.     , -3.23942,  2.59153]])
+    >>> X @ vr
+    array([[ 1.     ,  1.69593,  1.9246 ],
+           [ 0.     ,  2.59153,  3.23942],
+           [ 0.     , -3.23942,  2.59153]])
+    """
+    w1 = _asarray_validated(w)
+    v1 = _asarray_validated(v)
+    if len(w1.shape) < 1 or len(w1.shape) > 2:
+        raise ValueError('expected 1D array or stacked 1D arrays')
+    if (len(v1.shape) < 2 or len(v1.shape) > 3
+        or v1.shape[-2] != v1.shape[-1]
+        or v1.shape[-1] < 2):
+        raise ValueError('expected square matrix or stacked square matrices')
+    if len(v1.shape) != len(w1.shape) + 1 or v1.shape[-1] != w1.shape[-1]:
+        raise ValueError('expected eigenvalues and associated eigenvectors')
+
+    # number of stacked matrices and number of eigenvalues in each matrix
+    if w.ndim == 1:
+        stacked = False
+        M, n = 1, len(w)
+        w = w[numpy.newaxis, :]
+        v = v[numpy.newaxis, :, :]
+    else:
+        stacked = True
+        M, n = w.shape
+
+    # get indices for each first pair of complex eigenvalues
+    idx_im = numpy.argwhere(numpy.iscomplex(w))
+
+    # check if all complex eigenvalues have conjugate pairs
+    if len(idx_im[::2, 0]) != len(idx_im[1::2, 0]):
+        raise ValueError('expected complex-conjugate pairs of eigenvalues')
+
+    # all eigenvalues to diagonal form
+    w_mat = numpy.zeros((M, n, n))
+    di = range(n)
+    w_mat[:, di, di] = w[:, di].real
+
+    # complex eigenvalues to real block diagonal form
+    mat, k = idx_im[::2, 0], idx_im[::2, 1]
+    w_mat[mat, k, k+1] = w[mat, k].imag
+    w_mat[mat, k+1, k] = w[mat, k+1].imag
+    wr = w_mat.real
+
+    # compute real eigenvectors associated with real block diagonal eigenvalues
+    u = numpy.array([numpy.eye(n, n,
+                    dtype=numpy.complex64)] * M).reshape(M, n, n)
+    u[mat, k, k] = 0.5j
+    u[mat, k, k+1] = 0.5
+    u[mat, k+1, k] = -0.5j
+    u[mat, k+1, k+1] = 0.5
+    vr = numpy.matmul(v, u).real
```

Comment:
`(v @ u).real` would suffice

Reply:
We still support Python 2.7, so we can't use `@`.

Reply:
`matmul` was added to numpy version 1.10, but we support numpy back to version 1.8.2.

Reply:
Ah, my bad. I was still thinking about the docstrings.

Reply:
Does this mean that I cannot use matmul? Or should I check what version of numpy the user has, and then if python.version > 1.10 us matmul, and if python.version < 1.10 use for loops and dot?

Reply:
Only `v.dot(u).real` will suffice as it is supported by the newer versions anyway

Reply:
``v.dot(u).real`` would sufice only for single arrays, but for stacked arrays I either have to use matmul, or iterate over each matrix and multiply them by dot

Reply:
this would be a possible solution:
```    numpy_version = fromstring(numpy.__version__, dtype=int, sep='.')
    if numpy_version[0] >= 1 and numpy_version[1] >= 10:
        vr = numpy.matmul(v, u).real
    else:
        vr = zeros((M, n, n))
        for m in range(M):
            vr[m, :, :] = v[m, :, :].dot(u[m, :, :]).real
```

Reply:
You don't need to do these. You can use `einsum` or `tensordot` or manual reshaping or anything else. First hit of a quick search https://stackoverflow.com/questions/5344843/a-loopless-3d-matrix-multiplication-in-python but there are many other ways to do tensor product. I often fallback to einsum if the product gets a bit complicated and it is pretty well-optimized.

Reply:
Your're right, I used einsum and it works as fast as matmul for all cases.